Annotate LibWebRTC endpoints with feature flags

darryl-apple opened the pull request:
#### 46aa00cb99c202f77fb48e2c0f2e8ba137680704
<pre>
Annotate LibWebRTC endpoints with feature flags
<a href="https://bugs.webkit.org/show_bug.cgi?id=279969">https://bugs.webkit.org/show_bug.cgi?id=279969</a>
<a href="https://rdar.apple.com/problem/136286660">rdar://problem/136286660</a>

Reviewed by Youenn Fablet and Sihui Liu.

Annotate LibWebRTC endpoints with feature flags

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
(WebCore::RTCPeerConnection::create):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::configureLoggingChannel):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::updateSharedPreferencesForWebProcess):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::updateSharedPreferencesForWebProcess): Deleted.
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.h:
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.messages.in:
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
(WebKit::LibWebRTCCodecsProxy::create):
(WebKit::LibWebRTCCodecsProxy::LibWebRTCCodecsProxy):

Canonical link: <a href="https://commits.webkit.org/284463@main">https://commits.webkit.org/284463@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fb4c1ee92df70132f43f30ef8c15aea1b689638

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48864 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22137 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73547 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20622 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71581 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56665 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20473 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55241 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13703 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72530 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44571 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59974 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35720 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41238 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17404 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18999 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/62580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63179 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17750 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75258 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68710 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13446 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16970 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13486 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60057 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62817 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15421 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10840 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4457 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90492 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44668 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/16056 "Found 2 new JSC binary failures: testapi, testb3, Found 99 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Function/prototype.js.default, ChakraCore.yaml/ChakraCore/test/Operators/instanceof.js.default, jsc-layout-tests.yaml/js/script-tests/exception-instanceof.js.layout, jsc-layout-tests.yaml/js/script-tests/exception-instanceof.js.layout-dfg-eager-no-cjit, jsc-layout-tests.yaml/js/script-tests/exception-instanceof.js.layout-no-cjit, jsc-layout-tests.yaml/js/script-tests/function-apply-aliased.js.layout-no-cjit, jsc-layout-tests.yaml/js/script-tests/instance-of-immediates.js.layout, jsc-layout-tests.yaml/js/script-tests/instance-of-immediates.js.layout-dfg-eager-no-cjit, jsc-layout-tests.yaml/js/script-tests/instance-of-immediates.js.layout-no-cjit, jsc-layout-tests.yaml/js/script-tests/instanceof-operator.js.layout ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45742 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46937 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45483 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
```diff
@@ -49,6 +49,7 @@
 #import <wtf/Scope.h>
 #import <wtf/TZoneMallocInlines.h>
 
+
```
Not needed

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.h:
```diff
@@ -122,11 +122,14 @@ class LibWebRTCCodecsProxy final : public IPC::WorkQueueMessageReceiver {
 
     std::unique_ptr<WebCore::PixelBufferConformerCV> m_pixelBufferConformer;
+    ThreadSafeWeakPtr<GPUConnectionToWebProcess> m_gpuConnection;
+
```
Line not needed

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.h:
```diff
@@ -122,11 +122,14 @@ class LibWebRTCCodecsProxy final : public IPC::WorkQueueMessageReceiver {
     Ref<WorkQueue> m_queue;
     Ref<RemoteVideoFrameObjectHeap> m_videoFrameObjectHeap;
     WebCore::ProcessIdentity m_resourceOwner;
+    SharedPreferencesForWebProcess m_sharedPreferencesForWebProcess;
```
This is unused right now.

darryl-apple replied:
This is now used.

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
```diff
@@ -76,11 +77,18 @@
 
+const SharedPreferencesForWebProcess& LibWebRTCCodecsProxy::sharedPreferencesForWebProcess() const
+{
+    auto gpuConnectionToWebProcess = m_gpuConnection.get();
```
Can in theory be null.
I am not familiar with sharedPreferencesForWebProcess(), but here, it would probably be called from non main thread. Which might make this not really thread safe if it is accessed while being modified.

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.h:
```diff
@@ -67,14 +68,14 @@ class SharedVideoFrameReader;
 public:
-    static Ref<LibWebRTCCodecsProxy> create(GPUConnectionToWebProcess&);
+    static Ref<LibWebRTCCodecsProxy> create(GPUConnectionToWebProcess&, SharedPreferencesForWebProcess&);
     ~LibWebRTCCodecsProxy();
```
Should be SharedPreferencesForWebProcess& or SharedPreferencesForWebProcess

szewai replied:
Nit: `LibWebRTCCodecsProxy` does not need `GPUConnectionToWebProcess`, so the constructor can just take `Connection`. It seems better to avoid passing `GPUConnectionToWebProcess&` while `GPUConnectionToWebProcess` is being constructed.

darryl-apple replied:
Changed to be SharedPreferencesForWebProcess&.

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.h:
```diff
@@ -67,14 +68,14 @@ class SharedVideoFrameReader;
-
-private:
+    const SharedPreferencesForWebProcess& sharedPreferencesForWebProcess() const { return m_sharedPreferencesForWebProcess; }
+    void updateSharedPreferencesForWebProcess(SharedPreferencesForWebProcess sharedPreferencesForWebProcess) { m_sharedPreferencesForWebProcess = WTFMove(sharedPreferencesForWebProcess); }
```
Since we are writing in main thread and reading in queue, there is an inherent race.
I guess this is fine since for the security case, the parameters should be stable (or at least start very strict).

szewai replied:
I think it's more safe to make `m_sharedPreferencesForWebProcess` only accessed from `WorkQueue`, by dispatching task to `m_queue` for initialization and update.

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
```diff
@@ -64,18 +64,19 @@
     , m_queue(webProcessConnection.gpuProcess().libWebRTCCodecsQueue())
     , m_videoFrameObjectHeap(webProcessConnection.videoFrameObjectHeap())
     , m_resourceOwner(webProcessConnection.webProcessIdentity())
+    , m_sharedPreferencesForWebProcess(WTFMove(sharedPreferencesForWebProcess))
```
We should probably not move a &, although in that case this is only a bool structure.

szewai replied:
It's weird to move here as `parameters.sharedPreferencesForWebProcess` is moved again later in `GPUConnectionToWebProcess::GPUConnectionToWebProcess`.

darryl-apple committed:
Annotate LibWebRTC endpoints with feature flags
https://bugs.webkit.org/show_bug.cgi?id=279969
rdar://problem/136286660

Reviewed by Youenn Fablet and Sihui Liu.

Annotate LibWebRTC endpoints with feature flags

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
(WebCore::RTCPeerConnection::create):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::configureLoggingChannel):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::updateSharedPreferencesForWebProcess):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::updateSharedPreferencesForWebProcess): Deleted.
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.h:
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.messages.in:
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
(WebKit::LibWebRTCCodecsProxy::create):
(WebKit::LibWebRTCCodecsProxy::LibWebRTCCodecsProxy):

Canonical link: https://commits.webkit.org/284463@main
